tests/framework/scheduler.py

```diff
@@ -89,6 +91,26 @@ def pytest_runtest_logreport(self, report):
         """
         self._add_report(report)
 
+    @staticmethod
+    def deselect_if_marked(config, batch, marker_name):
+        """Deselect marked tests which are not explicitly selected."""
+        deselected = []
+        remaining = []
```

Comment:
I'm not sure we need this array. We can just remove from `batch[items]` if ?I understand correctly.

Reply:
You are right. I was under the impression that we can remove while iterating. Changed it as you suggested.

Reply:
Oh I was thinking at something like 

```
for i=0; i < items.len; i++ {
    if (...) {
        items.remove(i);
        i--;
    }
}
```

I'm not sure that `batch['items'].remove(item)` works correctly. Did you check it ? I found a thread on a forum stating something like:

```
>>> somelist = range(10)
>>> for x in somelist:
...     somelist.remove(x)
>>> somelist
[1, 3, 5, 7, 9]

>>> somelist = range(10)
>>> for x in somelist[:]:
...     somelist.remove(x)
>>> somelist
[]
```

So the result depends on the type of list that you're iterating on.

Reply:
> I was under the impression that we can't remove while iterating

We can remove while iterating, but it's not safe to do so.

When iterating over an array, the generator will keep track of the yielded element index. If you remove one element, the generator sequence may be mixed up.

For example, let's assume that we have:
```
arr = [1, 2, 3, 4]

for i in arr:
    print(i)
```

Running this snippet will print:
```
1
2
3
4
```

What if while looping, we remove one of the elements?

```
arr = [1, 2, 3, 4]

for i in arr:
    print(i)

    if i == 2:
        arr.remove(i)
```

Running this snippet will print:

```
1
2
4
```
Why wasn't `3` printed?
What happened here was that when element `2` was removed the list length decreased and the indexes of the elements following `2` were decremented:
```
 1   2   3   4 
[0] [1] [2] [3]
```
Became:
```
 1   3   4
[0] [1] [2]
```

Since at removal time the index was `1`, the next yielded element was `arr[2]`.

In order to remove elements while iterating one should either create a copy of that list or create a separate list of elements to be removed and remove them after the loop.

Reply:
Yes, you are right. I've checked it before for a list with a single element, which works. But having an example like yours it does not work anymore. I've added the sugar you mentioned above `[:]`. PTAL!